File: Page Replacement/OPT.c
#include <stdio.h>

int findOPT(int frames[], int pages[], int num_frames, int num_pages, int current) {
    int farthest = current, pos = -1, i, j;

    for (i = 0; i < num_frames; i++) {
        int found = 0;
        for (j = current + 1; j < num_pages; j++) {
            if (frames[i] == pages[j]) {
                if (j > farthest) {
                    farthest = j;
                    pos = i;
                }
                found = 1;
                break;
            }
        }
        if (!found) {  // If a page is not used in the future
            return i;
        }
    }

    return (pos == -1) ? 0 : pos;  // Default to the first frame if no farthest found
}

int main() {
    int frames[10], pages[30], num_frames, num_pages, i, j, pos, page_faults = 0;

    printf("Enter number of frames: ");
    scanf("%d", &num_frames);

    printf("Enter number of pages: ");
    scanf("%d", &num_pages);

    printf("Enter page reference string: ");
    for (i = 0; i < num_pages; i++)
        scanf("%d", &pages[i]);

    for (i = 0; i < num_frames; i++)
        frames[i] = -1;  // Initialize frames as empty

    printf("\nPage Reference | Frames");
    printf("\n---------------|------------------");

    for (i = 0; i < num_pages; i++) {
        int found = 0;

        // Check if the page is already in the frame
        for (j = 0; j < num_frames; j++) {
            if (frames[j] == pages[i]) {
                found = 1;
                break;
            }
        }

        if (!found) {  // Page fault occurs
            if (i < num_frames) {
                pos = i;  // First empty frame
            } else {
                pos = findOPT(frames, pages, num_frames, num_pages, i);
            }
            frames[pos] = pages[i];
            page_faults++;
        }

        // Print the table row
        printf("\n%12d |", pages[i]);
        for (j = 0; j < num_frames; j++) {
            if (frames[j] != -1)
                printf(" %d", frames[j]);
            else
                printf(" -");
        }
    }

    printf("\n\nTotal Page Faults = %d\n", page_faults);
    return 0;
}

File: pyos/paging/opt.py
def findOptimal(pages, n, frames, m, curr_index):
    farthest = -1
    index_to_replace = -1

    for i in range(m):
        found = False
        for j in range(curr_index + 1, n):
            if frames[i] == pages[j]:
                found = True
                if j > farthest:
                    farthest = j
                    index_to_replace = i
                break
        if not found:
            return i  # No future use of this frame, replace it

    return index_to_replace


def main():
    pages = [7, 0, 1, 2, 0, 3, 0, 4, 2, 3, 0, 3]
    n = len(pages)

    frames = [-1] * 3
    m = len(frames)

    page_faults = 0

    for i in range(n):
        page = pages[i]
        found = False

        for j in range(m):
            if frames[j] == page:
                found = True
                break

        if not found:
            page_faults += 1
            replace_index = findOptimal(pages, n, frames, m, i)
            frames[replace_index] = page

        print(f"Frame status after page {page}: {frames}")

    print(f"Total page faults: {page_faults}")


if __name__ == "__main__":
    main()